platforms/axplat-riscv64-qemu-virt/src/irq.rs

```diff
@@ -0,0 +1,147 @@
+    /// and [`S_TIMER`] represents timer interrupt. If its value is [`S_EXT`], it
+    /// means it is an external interrupt, and the real IRQ number needs to
+    /// be obtained from PLIC.
+    /// 2. If its highest bit is 0, it means it is an interrupt on the device side,
+    /// and its value is equal to the IRQ number provided by PLIC.
+    fn register(irq: usize, handler: IrqHandler) -> bool {
+        with_cause!(
+            irq,
+            @S_TIMER => TIMER_HANDLER.compare_exchange(core::ptr::null_mut(), handler as *mut _, Ordering::AcqRel, Ordering::Acquire).is_ok(),
+            @S_EXT => {
+                warn!("External IRQ should be got from PLIC, not scause");
+                false
+            },
+            @EX_IRQ => {
+                if IRQ_HANDLER_TABLE.register_handler(irq, handler) {
+                    Self::set_enable(irq, true);
+                    true
+                } else {
+                    warn!("register handler for External IRQ {} failed", irq);
+                    false
+                }
+            }
+        )
+    }
+
+    /// Unregisters the IRQ handler for the given IRQ.
+    ///
+    /// It also disables the IRQ if the unregistration succeeds. It returns the
+    /// existing handler if it is registered, `None` otherwise.
+    fn unregister(irq: usize) -> Option<IrqHandler> {
+        with_cause!(
+            irq,
+            @S_TIMER => {
+                let handler = TIMER_HANDLER.swap(core::ptr::null_mut(), Ordering::AcqRel);
+                if !handler.is_null() {
+                    Some(unsafe { core::mem::transmute::<*mut (), IrqHandler>(handler) })
+                } else {
+                    None
+                }
+            },
+            @S_EXT => {
+                warn!("External IRQ should be got from PLIC, not scause");
+                None
+            },
+            @EX_IRQ => IRQ_HANDLER_TABLE.unregister_handler(irq)
+        )
+    }
+
+    /// Handles the IRQ.
+    ///
+    /// It is called by the common interrupt handler. It should look up in the
+    /// IRQ handler table and calls the corresponding handler. If necessary, it
+    /// also acknowledges the interrupt controller after handling.
+    fn handle(irq: usize) {
+        with_cause!(
+            irq,
+            @S_TIMER => {
+                trace!("IRQ: timer");
+                let handler = TIMER_HANDLER.load(Ordering::Acquire);
+                if !handler.is_null() {
+                    // SAFETY: The handler is guaranteed to be a valid function pointer.
+                    unsafe { core::mem::transmute::<*mut (), IrqHandler>(handler)() };
+                }
+            },
+            @S_EXT => {
+                // TODO: get IRQ number from PLIC
+                warn!("Not implemented: External IRQ should be got from PLIC, not scause");
+            },
+            @EX_IRQ => {
+                if !IRQ_HANDLER_TABLE.handle(irq) {
```

Comment:
`S_EXT` will be passed as the `irq` argument when handling PLIC IRQs, move this code to `@S_EXT`.